internal/controller/runtime/fsm/runtime_fsm_apply_clusterrolebindings.go

```diff
@@ -0,0 +1,192 @@
+		return nil, err
+	}
+
+	return shootClientWithAdmin, nil
+}
+
+func isRBACUserKind(s rbacv1.Subject) bool {
+	return s.Kind == rbacv1.UserKind &&
+		s.APIGroup == rbacv1.GroupName
+}
+
+func getRemoved(crbs []rbacv1.ClusterRoleBinding, admins []string) (removed []rbacv1.ClusterRoleBinding) {
+	// iterate over cluster role bindings to find out removed administrators
+	for _, crb := range crbs {
+		if !labels.Set(crb.Labels).AsSelector().Matches(labels.Set(labelsClusterRoleBindings)) {
+			// cluster role binding is not controlled by KIM
+			continue
+		}
+
+		index := slices.IndexFunc(crb.Subjects, isRBACUserKind)
+		if index < 0 {
+			// cluster role binding does not contain user subject
+			continue
+		}
+
+		subjectUserName := crb.Subjects[index].Name
+		if slices.Contains(admins, subjectUserName) {
+			continue
+		}
+		// administrator was removed
+		removed = append(removed, crb)
+	}
+
+	return removed
+}
+
+//nolint:gochecknoglobals
+var newContainsAdmin = func(admin string) func(rbacv1.ClusterRoleBinding) bool {
+	return func(r rbacv1.ClusterRoleBinding) bool {
+		for _, subject := range r.Subjects {
+			if !isRBACUserKind(subject) || subject.Name != admin {
+				continue
+			}
+			// admin found
+			return true
+		}
+		// admin not found in the slice
+		return false
+	}
+}
+
+func getMissing(crbs []rbacv1.ClusterRoleBinding, admins []string) (missing []rbacv1.ClusterRoleBinding) {
+	for _, admin := range admins {
+		containsAdmin := newContainsAdmin(admin)
+		if slices.ContainsFunc(crbs, containsAdmin) {
+			continue
+		}
+		crb := toAdminClusterRoleBinding(admin)
+		missing = append(missing, crb)
+	}
+
+	return missing
+}
+
+func toAdminClusterRoleBinding(name string) rbacv1.ClusterRoleBinding {
+	return rbacv1.ClusterRoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "admin-",
+			Labels:       labelsClusterRoleBindings,
+		},
+		Subjects: []rbacv1.Subject{{
+			Kind:     rbacv1.UserKind,
+			Name:     name,
+			APIGroup: rbacv1.GroupName,
+		}},
+		RoleRef: rbacv1.RoleRef{
+			APIGroup: rbacv1.GroupName,
+			Kind:     "ClusterRole",
+			Name:     "cluster-admin",
+		},
+	}
+}
+
+//nolint:gochecknoglobals
+var newDelCRBs = func(ctx context.Context, shootClient client.Client, crbs []rbacv1.ClusterRoleBinding) func() error {
+	return func() error {
+		for _, crb := range crbs {
+			if err := shootClient.Delete(ctx, &crb); err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+}
+
+//nolint:gochecknoglobals
+var newAddCRBs = func(ctx context.Context, shootClient client.Client, crbs []rbacv1.ClusterRoleBinding) func() error {
+	return func() error {
+		for _, crb := range crbs {
+			if err := shootClient.Create(ctx, &crb); err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+}
+
+func updateCRBApplyFailed(rt *imv1.Runtime) {
+	rt.UpdateStatePending(
+		imv1.ConditionTypeRuntimeConfigured,
+		imv1.ConditionReasonConfigurationErr,
+		string(metav1.ConditionFalse),
+		"failed to update kubeconfig admin access",
```

Comment:
The error message is too generic, suggest to pass it as the parameter to make it more specific.

Reply:
The detailed message is visible in the logs of the controller, we need to make sure we will not leak or give to much details to confuse the user.